File: 백준/Silver/9012. 괄호/괄호.py
def vps(str):
    stack = []
    for char in str:
        if char == '(':
            stack.append(char)
        elif char == ')':
            if stack:
                stack.pop() 
            else:
                return "NO"
    return "YES" if not stack else "NO"
                

T = int(input()) 
result = []

for _ in range(T):
    s = input()
    result.append(vps(s))

print("\n".join(result))

File: 백준/Silver/9012. 괄호/괄호.js
const fs = require("fs");
const input = fs.readFileSync(0, "utf8").trim().split("\n");

const t = +input[0];

for (let i = 1; i <= t; i++) {
  let ps = input[i].trim();
  let stack = [];
  let valid = true;

  for (let ch of ps) {
    if (ch === "(") {
      stack.push(ch);
    } else {
      if (stack.length === 0) {
        valid = false;
        break;
      }
      stack.pop();
    }
  }

  console.log(valid && stack.length === 0 ? "YES" : "NO");
}
